### PR TITLE
Add functionality to get NN model files from CMT so we can track an…

### PR DESCRIPTION
**Pull request by ahiguera-mx**
**What is the problem / what does the code in this PR do**
Add functionality to get NN files from CMT to do a better bookkeeping of any change

**Can you briefly describe how it works?**
please see this notebook 
https://github.com/XENONnT/analysiscode/blob/master/corrections/posrec_wCMT_example.ipynb

**Can you give a minimal working example (or illustrate with a figure)?**
please see this notebook
https://github.com/XENONnT/analysiscode/blob/master/corrections/posrec_wCMT_example.ipynb

### CODE REVIEW

File: straxen/plugins/position_reconstruction.py [hunk 1]
```diff
@@ -108,12 +110,8 @@ def _get_model_file_name(self):
 
 @export
 @strax.takes_config(
-    strax.Option("file_mlp",
-                 help="Name of saved MLP model file in the aux file data base."
-                      "The file contains both structure and weights. Set to "
-                      "None to skip the computation of this plugin.",
-                 default="xnt_mlp_wfsim_20201214.tar.gz"
-                 )
+    strax.Option('mlp_model',
```
Comment (jorana) [inline straxen/plugins/position_reconstruction.py:113]: can you:

1. reinstate the None explanation? 
2. also for GCN and CNN.
3. Furthermore, we should add True to the tuple.
4. Set the default to ('mlp_model', 'ONLINE', True), we'd want to be using this from the start.

**Review by jorana — COMMENTED** [review]
@ahiguera-mx , many thanks for adding this! It's clean and effective. Thanks for putting this together so fast. @terliuk and @reallsx are probably also happy with this PR!

I have some comments below. They can be summarized mostly by:

1. Can we allow direct paths to be input, as well as allowing `None`'s as input. Those would bypass the `get_corrections.get_NN_file` function. The `None`'s are a hard requirement, the direct paths are just me thinking it keeps things a little easier during testing.
2. Add the defaults CMT maps now, in this PR
3. `ffil` is save but for `v1` we use `interpolate`, does that also work with strings

File: straxen/corrections_services.py [hunk 2]
```diff
@@ -231,12 +229,32 @@ def get_pmt_gains(self, run_id, model_type, global_version,
...
+    def get_NN_file(self, run_id, model_type, global_version='ONLINE'):
+        """
+        Smart logic to return NN weights file name to be downloader by 
+        straxen.MongoDownloader()
+        :param run_id: run id from runDB
+        :param model_type: model type and neural network type; mlp, gcn, cnn 
+        :param global_version: global version
+        :param return: NN weights file name
+        """
+        if 'NN_weights' in model_type:
```
[the remark this answers is not in the document]
Comment (jorana) [reply straxen/corrections_services.py:241]: okay, I think it should be 'mlp_model' etc but the point remains the same. If its indeed 'mlp_model', most likely we should update the docstring

File: straxen/corrections_services.py [hunk 3]
```diff
@@ -76,6 +76,8 @@ def get_corrections_config(self, run_id, config_model=None):
             return self.get_pmt_gains(run_id, model_type, global_version)
         elif 'elife' in model_type:
             return self.get_elife(run_id, model_type, global_version)
+        elif 'NN_weights' in model_type:
+            return self.get_NN_file(run_id, model_type, global_version)
         else:
             raise ValueError(f'{correction} not found')
```
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py:82]: good catch!

File: straxen/corrections_services.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py:241]: great suggestion!

File: straxen/get_corrections.py [hunk 4]
```diff
@@ -109,3 +109,24 @@ def get_elife(run_id, elife_conf):
...
+        raise ValueError(f'NN_model must have three elements: '
+                         f'the model type, its specific configuration '
+                         f'and detector (True = nT)')
+    model_type, model_conf, is_nt = nn_conf
+    if model_type == 'CMT_model':
+        cmt = straxen.CorrectionsManagementServices(is_nt=is_nt)
+        nn_file = cmt.get_corrections_config(run_id, model_conf)
+        nn_file = ' '.join(map(str, nn_file))    
+
+    elif model_type == 'local':
```
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/get_corrections.py:126]: agree

File: straxen/plugins/position_reconstruction.py [hunk 5]
```diff
@@ -98,8 +99,9 @@ def compute(self, peaks):
         return result
 
     def _get_model_file_name(self):
-        config_file = f'file_{self.algorithm}'
-        model_file = self.config.get(config_file, "No file")
+        
+        config_file = f'{self.algorithm}_model'
```
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/plugins/position_reconstruction.py:103]: Indeed this is better way, I have implemented it

File: straxen/plugins/position_reconstruction.py (hunk 1 above)
Comment (ahiguera-mx) [reply straxen/plugins/position_reconstruction.py:121]: done

File: straxen/corrections_services.py [hunk 6]
```diff
@@ -231,12 +229,32 @@ def get_pmt_gains(self, run_id, model_type, global_version,
...
+        if 'NN_weights' in model_type:
+            correction = None
+            if 'mlp' in model_type:
+                correction = 'mlp'
+            elif 'cnn' in model_type:
+                correction = 'cnn'
+            elif 'gcn' in model_type:
+                correction = 'gcn'
+
+            file_name = self._get_correction(run_id, correction, global_version)
```
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py]: Hi @jorana 
Indeed, `v1` fails as with the following error 
`TypeError: Cannot interpolate with all object-dtype columns in the DataFrame. Try setting at least one column to a numeric dtype.` 
I need to think how to handle this, suggestions are welcome

[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py]: Hi @jorana 
I add a condition for `nn models` to use`fill` only, so this fixes the issue

**Comment by ahiguera-mx** [comment]
Thanks @jorana for the useful review, I've addressed your suggestions

File: straxen/corrections_services.py [hunk 7]
```diff
@@ -105,16 +107,16 @@ def _get_correction(self, run_id, correction, global_version,
                         # on when something was processed therefore
                         # don't interpolate but forward fill.
                         df = self.interface.interpolate(df, when, how='fill')
+                    if correction in ('mlp_model', 'cnn_model', 'gcn_model'):
```
Comment (jorana) [inline straxen/corrections_services.py:110]: I could also imagine that you want to add this to the `_get_corrections` function as an optional argument and then pass it from get_NN_file. This also works and is very clear.

File: straxen/get_corrections.py [hunk 8]
```diff
@@ -109,3 +109,22 @@ def get_elife(run_id, elife_conf):
...
+        raise ValueError(f'NN_model must be a tuple')
+    if not len(nn_conf) == 3:
+        raise ValueError(f'NN_model must have three elements: '
+                         f'the model type, its specific configuration '
+                         f'and detector (True = nT)')
+    model_type, model_conf, is_nt = nn_conf
+    if model_type == 'CMT_model':
+        cmt = straxen.CorrectionsManagementServices(is_nt=is_nt)
+        nn_file = cmt.get_corrections_config(run_id, model_conf)
+        nn_file = ' '.join(map(str, nn_file))
```
Comment (jorana) [inline straxen/get_corrections.py:124]: Sorry, this escaped my attention before, why do we need this line?

File: straxen/plugins/position_reconstruction.py [hunk 9]
```diff
@@ -98,22 +99,31 @@ def compute(self, peaks):
...
 @strax.takes_config(
-    strax.Option("file_mlp",
-                 help="Name of saved MLP model file in the aux file data base."
-                      "The file contains both structure and weights. Set to "
-                      "None to skip the computation of this plugin.",
-                 default="xnt_mlp_wfsim_20201214.tar.gz"
-                 )
+    strax.Option('mlp_model',
+                 help='Neural network model.' 
+                      'If CMT, specify as (CMT_model, (mlp_NN_weights, ONLINE), True)))'
```
Comment (jorana) [inline straxen/plugins/position_reconstruction.py:123]: A little too fast a copy pasta 😄 , the help is now the same for all of the plugins.

**Review by jorana — APPROVED** [review]
Many thanks again for the changes! Works like a charm 😄 :
```python
>>> import straxen
>>> straxen.get_NN_file('0', ("CMT_model", ('mlp_model', "ONLINE"), True))
'xnt_mlp_wfsim_20201214.tar.gz'
>>> straxen.get_NN_file('0', ("CMT_model", ('gcn_model', "ONLINE"), True))
'xnt_gcn_wfsim_20201203.tar.gz'
>>> straxen.get_NN_file('0', ("CMT_model", ('cnn_model', "ONLINE"), True))
'xnt_cnn_wfsim_A_5_2000_20210112.tar.gz'
```

File: straxen/plugins/position_reconstruction.py (hunk 9 above)
Comment (ahiguera-mx) [reply straxen/plugins/position_reconstruction.py:123]: yeah too much copy "pasta"... that sounds good

File: straxen/get_corrections.py (hunk 8 above)
Comment (ahiguera-mx) [reply straxen/get_corrections.py:124]: not sure which line you are referring, but if is the last line, it is just for convince when I save the models into pandasDataFrame, and make sure if returns a simple string

File: straxen/corrections_services.py (hunk 7 above)
Comment (ahiguera-mx) [reply straxen/corrections_services.py:110]: we could do that, lets see once we have more corrections, then we can have a better idea how we can improve the functionality of  `_get_corrections`

File: straxen/get_corrections.py (hunk 8 above)
Comment (jorana) [reply straxen/get_corrections.py:124]: okay, `str(nn_file)` wasn't enough? (just curious doesn't matter at all)

**Review by jorana — APPROVED** [review]

**Review by WenzDaniel — APPROVED** [review]
Nothing to add. Thanks Aaron

**Pull request merged**